metals/src/main/scala/scala/meta/internal/bsp/ConnectionBspStatus.scala

```diff
@@ -13,25 +16,83 @@ class ConnectionBspStatus(
     bspStatus: BspStatus,
     folderPath: AbsolutePath,
     icons: Icons,
-) {
-  private val isServerResponsive = new AtomicBoolean(false)
-  val status: MetalsStatusParams => Unit = bspStatus.status(folderPath, _)
-
-  def connected(serverName: String): Unit =
-    if (isServerResponsive.compareAndSet(false, true))
-      status(ConnectionBspStatus.connectedParams(serverName, icons))
-  def noResponse(serverName: String): Unit =
-    if (isServerResponsive.compareAndSet(true, false)) {
-      scribe.debug("server liveness monitor detected no response")
-      status(ConnectionBspStatus.noResponseParams(serverName, icons))
-    }
+)(implicit rc: ReportContext) {
+  private val status = new AtomicReference[BspStatusState](
+    BspStatusState(Disconnected, None, None)
+  )
+  private val currentSessionErrors = new AtomicReference[Set[String]](Set())
+
+  def connected(serverName: String): Unit = changeState(Connected(serverName))
+
+  def noResponse(): Unit = {
+    scribe.debug("server liveness monitor detected no response")
+    changeState(NoResponse)
+  }
 
   def disconnected(): Unit = {
-    isServerResponsive.set(false)
-    status(ConnectionBspStatus.disconnectedParams)
+    currentSessionErrors.set(Set.empty)
+    changeState(Disconnected)
+  }
+
+  def showError(message: String, pathToReport: Path): Unit = {
+    val updatedSet =
+      currentSessionErrors.updateAndGet(_ + pathToReport.toUri().toString())
+    changeState(ErrorMessage(message), updatedSet)
+  }
+
+  def onReportsUpdate(): Unit = {
```

Comment:
Just to make sure, we don't want to recompute `currentSessionErrors` after a report is deleted?

Reply:
`showState` always takes an intersection of errors in `currentSessionErrors` and reports, I don't care to remove the deleted ones, because I don't think it matters. But if you think it would be more comprehensive I can count the intersection and update `currentSessionErrors` here.

Reply:
No, i think it's fine the way it is